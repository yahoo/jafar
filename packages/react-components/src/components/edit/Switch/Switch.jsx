/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import InternalSwitch from '@material-ui/core/Switch';

/**
 * Represent a boolean value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Switch/Switch.jsx">
 Switch</a> from '@jafar-org/react-components/edit/Switch'
 */
export default class Switch extends React.Component {
  static propTypes = {
    value: PropTypes.bool,
    disabled: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: false,
    disabled: false,
  };

  render() {
    return (
      <InternalSwitch
        onChange={this.onValueChange}
        checked={this.props.value}
        disabled={this.props.disabled} />
    );
  }
 
 onValueChange = (e) => {
   this.props.onValueChange(e.target.checked);
 }
}
