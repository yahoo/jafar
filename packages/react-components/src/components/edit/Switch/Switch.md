<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    status: {
      label: 'Status',
      path: 'status',
      component: {
        name: 'Switch',
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      status: true, // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    Switch: {
      renderer: Switch,
    }, 
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='status' />
</Form>
```


<h4>Simple</h4>

```javascript
initialState = { 
    status: true,
};

 <Switch
    value={state.status}
    onValueChange={(status) => {
        setState({ status });
    }}
    />
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
    status: undefined,
};
 <Switch
    value={state.status}
    onValueChange={(status) => {
        setState({ status });
    }}
    />
```

<h4>Disabled</h4>

```javascript
initialState = { 
    status: true,
};
 <Switch
    value={state.status}
    disabled={true}
    onValueChange={(status) => {
        setState({ status });
    }}
    />
```
