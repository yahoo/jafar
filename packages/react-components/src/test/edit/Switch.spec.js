/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import Switch from '../../components/edit/Switch/index.js';
import { mapper } from '../../components/edit/Switch/Switch';

describe('Switch', () => {
  let componentProps;

  const jafarProps = {
    value: true,
    disabled: true,
    onValueChange: jest.fn(),
  };

  const expectedProps = {
    checked: true,
    disabled: true,
    onChange: expect.any(Function),
  };
 
  beforeEach(() => {
    componentProps = mapper(jafarProps);
  });

  describe('mapper', () => {
    it('mapper return correct props', () => {
      expect(componentProps).toEqual(expectedProps);
    });
  
    it('call onValueChange with correct value', () => {
      const mockEvent = { target: { checked: false } };
      componentProps.onChange(mockEvent);
      expect(jafarProps.onValueChange).toHaveBeenCalledWith(false);
    });
  });

  describe('component', () => {
    it('renders ok', () => {
      const component = shallow(<Switch {...jafarProps} />);
      expect(component.props()).toEqual(expectedProps);
    });
  });
});
