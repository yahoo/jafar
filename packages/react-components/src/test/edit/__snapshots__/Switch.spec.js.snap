// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should render provided data 1`] = `
<WithStyles(ForwardRef(Switch))
  checked={true}
  disabled={false}
  onChange={[Function]}
/>
`;
